grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcServiceImplErrorTest.java

```diff
@@ -0,0 +1,303 @@
+            final ServiceRequestContextCaptor serviceCaptor = server.requestContextCaptor();
+            assertThat(serviceCaptor.size()).isEqualTo(1);
+            final ServiceRequestContext serviceCtx = serviceCaptor.take();
+            assertThat(serviceCtx.log().ensureComplete().responseCause())
+                    .isInstanceOf(StatusRuntimeException.class)
+                    .satisfies(cause -> {
+                        assertThat(Status.fromThrowable(cause).getCode()).isEqualTo(
+                                Status.ABORTED.getCode());
+                        assertThat(Status.trailersFromThrowable(cause)).satisfies(metadata -> {
+                            // metadata.keys() throws IndexOutOfBoundsException as metadata is corrupted.
+                            assertThat(metadata.containsKey(
+                                    InternalMetadata.keyOf(KEY_OF_CORRUPTED_METADATA,
+                                                           Metadata.ASCII_STRING_MARSHALLER)
+                            )).isTrue();
+                        });
+                    });
+        }
+    }
+
+    // Normal case of #onError at Unary RPC.
+    @Test
+    void clientUnaryCall2() throws InterruptedException {
+        try (ClientRequestContextCaptor captor = Clients.newContextCaptor()) {
+            assertThatThrownBy(() -> {
+                final TestServiceBlockingStub client =
+                        GrpcClients.newClient("http://127.0.0.1:" + server.httpPort(),
+                                              TestServiceBlockingStub.class);
+                client.unaryCall2(REQUEST_MESSAGE);
+            }).satisfies(cause -> {
+                assertThat(Status.fromThrowable(cause).getCode()).isEqualTo(Status.ABORTED.getCode());
+                assertThat(Status.trailersFromThrowable(cause)).satisfies(metadata -> {
+                    for (String key : validMetadata.keys()) {
+                        final Key<String> metaKey = Key.of(key, Metadata.ASCII_STRING_MARSHALLER);
+                        assertThat(metadata.get(metaKey)).isEqualTo(validMetadata.get(metaKey));
+                    }
+                });
+            });
+
+            final RequestLog log = captor.get().log().whenComplete().join();
+            assertThat(log.responseStatus()).isEqualTo(HttpStatus.OK);
+            assertThat(log.responseTrailers().get(GrpcHeaderNames.GRPC_STATUS)).isNull();
+            assertThat(log.responseHeaders().get(GrpcHeaderNames.GRPC_STATUS)).satisfies(grpcStatus -> {
+                assertThat(grpcStatus).isNotNull();
+                assertThat(grpcStatus).isEqualTo(String.valueOf(Status.ABORTED.getCode().value()));
+            });
+            assertThat(log.responseCause())
+                    .isInstanceOf(StatusRuntimeException.class)
+                    .satisfies(cause -> {
+                        assertThat(Status.fromThrowable(cause).getCode()).isEqualTo(Status.ABORTED.getCode());
```

Comment:
```suggestion
                        assertThat(Status.fromThrowable(cause).getCode()).isEqualTo(Status.INTERNAL.getCode());
```

I think this probably needs to be `INTERNAL` since that was what was sent from server-side.
To do this, I guess we can't get away with catching the exception in `AbstractServerCall#statusToTrailers` and we need to catch the exception in the upper layer.

What do you think of moving the exception catching logic up to `AbstractServerCall#onClose`? I've drafted a simple poc on what I think might be a potential fix. Sorry about the direction change 😅 

https://github.com/jrhee17/armeria/commit/117339f2d9bf2e4f4dd37319d26ec2b7c22f1f62

Reply:
No problem.

If you are alright about the method signatures, etc in https://github.com/aki-s/armeria/pull/1/commits
,
I wish to merge your changes

Reply:
Alright, I've merged it.

Reply:
`unaryCall2()` returns valid metadata with ABORT, so this line doesn't have to change.